descs: push descriptor type hydration into the desc.Collection

rohany opened the pull request:
Fixes #49484.
Preparation for #51385.

Up until now, the `planner` was responsible for installing user defined
type metadata in tables that contained user defined types. This was
slightly messy and caused leakage of responsibility regarding when
descriptors had user defined types vs when they didn't. This commit
pushes that responsibility into the `descs.Collection`. It also paves
the way for work to avoid copying `ImmutableTableDescriptor`s that
contain user defined types every time that they need hydration.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51621)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> Reviewed 3 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51621)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @rohany)

___
*[pkg/sql/catalog/descs/collection.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51621#-MClf6t04wk1EX4YQneG:-MClf6t14_D1IvpRMlIS:btq007u) ([raw file](https://github.com/cockroachdb/cockroach/blob/91a0c1b2286fa7209c941fe2a30afe9bdfdd3d47/pkg/sql/catalog/descs/collection.go#L263)):*
> ```Go
> 		// not shared.
> 		return desc, sqlbase.HydrateTypesInTableDescriptor(ctx, t.TableDesc(), sqlbase.TypeLookupFunc(getType))
> 	case *sqlbase.ImmutableTableDescriptor:
> ```

How does one get a handle to an ImmutableTableDescriptor that isn't hydrated?


<!-- Sent from Reviewable.io -->

rohany commented:
> How does one get a handle to an ImmutableTableDescriptor that isn't hydrated?

The lease manager/uncached accessor is pulling up table descs and creating immutable descs to hand out to higher levels of name resolution -- those aren't coming hydrated.

ajwerner reviewed (commented):
>The lease manager/uncached accessor is pulling up table descs and creating immutable descs to hand out to higher levels of name resolution (i.e. the collection) -- those aren't coming hydrated.

That's sort of exactly my question. Is the `Collection` the only thing that ever has handles to unhydrated `ImmutableTableDescriptor`s? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51621#-:-MClkG19AOJElrsLKVKd:bbbup4d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @rohany)



<!-- Sent from Reviewable.io -->

rohany commented:
Yeah, I think that is a reasonable separation -- everything below the collection needs to hydrate descriptors if they want to use them.

rohany committed:
descs: push descriptor type hydration into the desc.Collection

Fixes #49484.
Preparation for #51385.

Up until now, the planner was responsible for installing user defined
type metadata in tables that contained user defined types. This was
slightly messy and caused leakage of responsibility regarding when
descriptors had user defined types vs when they didn't. This commit
pushes that responsibility into the descs.Collection. It also paves
the way for work to avoid copying ImmutableTableDescriptors that
contain user defined types every time that they need hydration.

Release note: None

rohany commented:
Rebased, ready for review.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51621#-:-MCls7lp-3UD4fLCwY2o:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang and @rohany)



<!-- Sent from Reviewable.io -->

rohany commented:
bors r+
